### PR TITLE
DM-22069: Add ProductBoundedField.

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: tests/test_chebyshevBoundedField.py [hunk 1]
```diff
@@ -90,6 +90,8 @@ def setUp(self):
 
         array = np.arange(self.bbox.getArea(), dtype=np.float32).reshape(self.bbox.getDimensions())
         self.image = lsst.afw.image.ImageF(array)
+        self.fields = [lsst.afw.math.ChebyshevBoundedField(self.bbox, coeffs) for _, coeffs in self.cases]
+        self.product = lsst.afw.math.ProductBoundedField(self.fields)
```
Comment (erykoff) [inline tests/test_chebyshevBoundedField.py]: Should there be an additional test product where these are done in the opposite order, just to confirm commutitivity, or does that go without saying?

File: include/lsst/afw/math/ProductBoundedField.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ *  A BoundedField that lazily multiplies a sequence of other BoundedFields.
+ */
+class ProductBoundedField final : public table::io::PersistableFacade<ProductBoundedField>,
+                                  public BoundedField {
+public:
+
+    /**
+     *  Construct from a sequence of nested BoundedFields.
+     *
+     *  @param[in] nested  BoundedFields to be multiplied together.
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply include/lsst/afw/math/ProductBoundedField.h]: I think of these docstring comments as being aligned, whitespace-delimited columns, but I guess there's no point to use more than one space when there's only one row.

File: include/lsst/afw/math/ProductBoundedField.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ *  A BoundedField that lazily multiplies a sequence of other BoundedFields.
+ */
+class ProductBoundedField final : public table::io::PersistableFacade<ProductBoundedField>,
+                                  public BoundedField {
+public:
+
+    /**
+     *  Construct from a sequence of nested BoundedFields.
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply include/lsst/afw/math/ProductBoundedField.h]: I don't think it's really a term of art.  Would "factor" make more sense, since that's the formal name for a term in a product?

[the remark this answers is not in the document]
Comment (erykoff) [reply include/lsst/afw/math/ProductBoundedField.h]: Yes, that does make more sense to me.

File: tests/test_chebyshevBoundedField.py (hunk 1 above)
Comment (TallJimbo) [reply tests/test_chebyshevBoundedField.py]: I'm not really worried about commutivity except at the level of floating point math edge cases, and I don't think it's worth trying to deal with those.
